src/storage/v2/schema_validator.cpp

```diff
@@ -0,0 +1,107 @@
+// included in the file licenses/BSL.txt; by using this file, you agree to be bound by the terms of the Business Source
+// License, and you may not use this file except in compliance with the Business Source License.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+#include "storage/v2/schema_validator.hpp"
+
+#include <bits/ranges_algo.h>
+#include <cstddef>
+#include <ranges>
+
+#include "storage/v2/schemas.hpp"
+
+namespace memgraph::storage {
+
+bool operator==(const SchemaViolation &lhs, const SchemaViolation &rhs) {
+  return lhs.status == rhs.status && lhs.label == rhs.label &&
+         lhs.violated_schema_property == rhs.violated_schema_property &&
+         lhs.violated_property_value == rhs.violated_property_value;
+}
+
+SchemaViolation::SchemaViolation(ValidationStatus status, LabelId label) : status{status}, label{label} {}
+
+SchemaViolation::SchemaViolation(ValidationStatus status, LabelId label, SchemaProperty violated_schema_property)
+    : status{status}, label{label}, violated_schema_property{violated_schema_property} {}
+
+SchemaViolation::SchemaViolation(ValidationStatus status, LabelId label, SchemaProperty violated_schema_property,
+                                 PropertyValue violated_property_value)
+    : status{status},
+      label{label},
+      violated_schema_property{violated_schema_property},
+      violated_property_value{violated_property_value} {}
+
+SchemaValidator::SchemaValidator(Schemas &schemas) : schemas_{schemas} {}
+
+[[nodiscard]] std::optional<SchemaViolation> SchemaValidator::ValidateVertexCreate(
+    LabelId primary_label, const std::vector<LabelId> &labels,
+    const std::vector<std::pair<PropertyId, PropertyValue>> &properties) const {
+  // Schema on primary label
+  const auto schema = schemas_.GetSchema(primary_label);
+  if (!schema) {
+    return SchemaViolation(SchemaViolation::ValidationStatus::NO_SCHEMA_DEFINED_FOR_LABEL, primary_label);
+  }
+
+  // Is there another primary label among secondary labels
+  for (const auto &secondary_label : labels) {
+    if (schemas_.GetSchema(secondary_label) != std::nullopt) {
+      return SchemaViolation(SchemaViolation::ValidationStatus::VERTEX_SECONDARY_LABEL_IS_PRIMARY, secondary_label);
+    }
+  }
+
+  // Check only properties defined by schema
+  for (const auto &schema_type : schema->second) {
+    // Check schema property existence
+    auto property_pair = std::ranges::find_if(
+        properties, [schema_property_id = schema_type.property_id](const auto &property_type_value) {
+          return property_type_value.first == schema_property_id;
+        });
+    if (property_pair == properties.end()) {
+      return SchemaViolation(SchemaViolation::ValidationStatus::VERTEX_HAS_NO_PROPERTY, primary_label, schema_type);
```

Comment:
```
      return SchemaViolation(SchemaViolation::ValidationStatus::VERTEX_HAS_NO_PRIMARY_PROPERTY, primary_label, schema_type);
```

or 

```
      return SchemaViolation(SchemaViolation::ValidationStatus::SCHEMA_HAS_NO_PROPERTY, primary_label, schema_type);
```

Would be a better name in my opinion.

Reply:
I will for the first one, since all status already have vertex suffix